### DISCUSSION TITLE
iKonvert serial to json

### DISCUSSION THREAD

**Discussion by anilbhatt1** [canboat/canboat#286] [Q&A] [answered]
I am planning to use an iKonvert NMEA2000 to USB converter. Purpose is to get NMEA2000 message format in json format. I will plug iKonvert to my linux based pc and  json format will be fed to another program running on same linux PC.Will below CLI command executed from my linux terminal get me json format : 

    ikonvert-serial -r /dev/ttyS1 |  analyzer -json

PS: I know this is not an issue but a query. Posting this in issues as I have noticed very less discussions (4) compared to issues logged (120+)

**Comment by sbender9** [comment]
Check out the Signal K server.

**Reply by anilbhatt1** [reply]
Thx @sbender9 ...checking on that

**Comment by keesverruijt** [comment]
Yes.

**Comment by anilbhatt1** [comment]
Thx ! Is there a command or way to convert ikonvert.log (in samples) to json format. For example, below command worked well.
**analyzer -raw < ./samples/126208.raw** 
But, when I tried to convert ikonvert.log in samples to json format via analyzer via below CLI command (Ubuntu 22.04)
**analyzer -json < ./samples/ikonvert.log** 
I am getting errors as below. 
_ERROR 2022-09-08T07:11:41.536Z [analyzer] Error reading message, scanned 29 bytes from !PDGY,127251,2,3,255,798.565,VFDk6/////8=
/VFDk6/////8=
, index 0ERROR 2022-09-08T07:11:41.536Z [analyzer] Unknown message error 2: !PDGY,127251,2,3,255,798.565,VFDk6/////8=_
Please advise.

**Reply by keesverruijt** [reply]
This is because the ikonvert.log is a sample of what you find on the serial port. It is not in the CANboat raw format which is shared by most of the tools.

If you pass this as a filename to `ikonvert-serial` it will understand it is not a serial port and convert the output to canboat raw format. Try something like:

    ikonvert-serial ./samples/ikonvert.log | analyzer -json

**Reply by anilbhatt1** [reply]
Thx @keesverruijt ! Now, I am able to get it in json format. Kindly confirm if my understanding below is correct.
If I connect ikonvert to my Linux PC and give below command:

    ikonvert-serial -r /dev/ttyS1 |  analyzer -json

then

  `ikonvert-serial -r /dev/ttyS1` -> ikonvert-serial will convert NMEA2000 to iKonvert format which is like `!PDGY,129039,4,43,255,481.706,ErZcAw6BGlf/h2BIHkP//wAABgAG//8AdAH/`
and then
`analyzer -json` -> analyzer will accept this iKonvert format and convert it to canboat json format like `{"timestamp":"2022-09-15T13:45:21.595Z","prio":2,"src":3,"dst":255,"pgn":127251,"description":"Rate of Turn","fields":{"SID":79,"Rate":-3.07764}}`

**Reply by keesverruijt** [reply]
Nope. You are wrong. The !PDGY format is what is on the USB serial port. See https://github.com/digitalyacht/iKonvert. `ikon vert-serial` converts this format to and from canboat-raw.

In ASCII art:

```
   NMEA 2000 CAN bus
        |
        v
       CAN
   iKonvert
       USB
        |
        | !PDGY,129039,4,43,255,481.706,ErZcAw6BGlf/h2BIHkP//wAABgAG//8AdAH/
        |
        v
       USB
    ikonvert-serial
        |
        | 2022-09-15T14:17:14.564Z,4,129039,43,255,27,12,b6,5c,03,0e,81,1a,57,ff,87,60,48,1e,43,ff,ff,00,00,06,00,06,ff,ff,00,74,01,ff
        |
        v
    analyzer -json
        |
        \> {"timestamp":"2022-09-15T14:17:37.845Z","prio":4,"src":43,"dst":255,"pgn":129039,
           "description":"AIS Class B Position Report",
           "fields":{"Message ID":18,"Repeat Indicator":"Initial","User ID":"235101366",...
```

Here is another example using the `ikonvert.log` file:

    $ head -1 samples/ikonvert.log
    !PDGY,129039,4,43,255,481.706,ErZcAw6BGlf/h2BIHkP//wAABgAG//8AdAH/
    $ ikonvert-serial samples/ikonvert.log | head -1
    INFO 2022-09-15T14:17:14.563Z [ikonvert-serial] Device is a normal file, do not set the attributes.
    2022-09-15T14:17:14.564Z,4,129039,43,255,27,12,b6,5c,03,0e,81,1a,57,ff,87,60,48,1e,43,ff,ff,00,00,06,00,06,ff,ff,00,74,01,ff
    $ ikonvert-serial samples/ikonvert.log | head -1 | analyzer
    {"timestamp":"2022-09-15T14:17:37.845Z","prio":4,"src":43,"dst":255,"pgn":129039,"description":"AIS Class B Position Report","fields":{"Message ID":18,"Repeat Indicator":"Initial","User ID":"235101366","Longitude":-1.10688,"Latitude":50.80598,"Position Accuracy":"High","RAIM":"in use","Time Stamp":"16","SOG":0.00,"AIS Transceiver information":"Channel A VDL reception","Regional Application":0,"Regional Application":0,"Unit type":"CS","Integrated Display":"No","DSC":"Yes","Band":"Entire marine band","Can handle Msg 22":"Yes","AIS mode":"Autonomous","AIS communication state":"ITDMA"}}

**Reply by keesverruijt** [reply]
Notice how the !PDGY format is what is on the serial line (USB). The `ikonvert-serial` program convers this to the format starting with a timestamp and encoding all bytes in 2 byte hex digits.

**Reply by anilbhatt1** [reply]
Got it ! Thx @keesverruijt for detailed explanation !

**Comment by anilbhatt1** [comment]
I was trying to understand the ikonvert-serial code using linux terminal.  
I am new to C (knows C++ concepts but didnt code much). Hence to confirm my understanding, I was trying to put printf statements inside ikonvert-serial.c file and then compiling via make command in linux. Intention was to get these statements printed in terminal next time when I run pgm.
But, the changes I was making in code was not getting reflected in terminal. I was under the impression that executable resides in rel/linux-x86_64 folder. 
Inorder to confirm that, I deleted the executable and then executed the command "ikonvert-serial ./samples/ikonvert.log". I was expecting an error. But, program ran without any issues as earlier.
Can you suggest a way to debug the code ?

@keesverruijt  I know this is a silly question to ask. But, thought of asking rather than searching in dark.

**Reply by anilbhatt1** [reply]
Never mind...I figured it out. Executables were lying in **/usr/local/bin** as a result of _"sudo make install_" command.

**Comment by anilbhatt1** [comment]
@keesverruijt 
Kindly advise.
Read this in https://signalk.org/installation.html. "_This software called [CANboat](https://github.com/canboat/canboatjs) is a suite of tools that can read and **write NMEA 2000 data and convert it in to Signal K._**"
Can we feed analyzer (C module) json output as shown below to Signal-K server ? 
_{"timestamp":"2022-10-17T15:17:51.935Z","prio":2,"src":3,"dst":255,"pgn":127250,"description":"Vessel Heading","fields":{"SID":242,"Heading":264.3,"Deviation":0.0,"Variation":0.0,"Reference":"Magnetic"}}_

**Reply by tkurki** [reply]
Answered in https://github.com/SignalK/signalk-server/issues/1485

**Reply by anilbhatt1** [reply]
Yeah seen that. Thx @tkurki

**Comment by anilbhatt1** [comment]
Hi,

Is there any program available to convert NMEA2K data to NMEA0183 format and write to a file ?

**Reply by keesverruijt** [reply]
`n2kd --nmea0183` will translate the most common messages to NMEA0183.

**Reply by anilbhatt1** [reply]
@keesverruijt Is below CLI command usage correct ? I am trying to write the nmea0183 output to out1.txt.

ikonvert-serial ./ikonvert.log | analyzer -json | n2kd --nmea0183 > ./out1.txt

**Comment by keesverruijt** [comment, marked as answer]
The current version of n2kd requires that you feed it with `analyzer -json -nv`. Look at the `n2kd/tests/Makefile` for various examples.
